### PR TITLE
POSIX signal handling

### LINKED ISSUE #770: Linux signals.
@marksamman I know I mentioned this before and I said I would figure it out or whatever but could you please add support for them? Just basic ones like reload all and save and quit.

I tried doing it myself but I have not been able  to do it successfully(I'm not a programmer).

Just the "Save & Quit" signal would be a huge blessing.

Thanks!
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1924029-linux-signals?utm_campaign=plugin&utm_content=tracker%2F392151&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F392151&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by djarek**
Closes #770 
Added signal handling for SIGTERM, SIGHUP, SIGINT, SIGUSR1. The process of receiving signals is done throught asio's signal_set which allows us to uniformly handle both signals an Windows console events.

### CODE REVIEW

**Comment by Nu77** [comment]
Nice :+1:

File: src/signalhandlingthread.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+extern Chat* g_chat;
+extern LuaEnvironment g_luaEnvironment;
+
+void SignalHandlingThread::start()
+{
+	setState(THREAD_STATE_RUNNING);
+	thread = std::thread(&SignalHandlingThread::signalHandlingThread, this);
+}
+
+sigset_t SignalHandlingThread::setupSignalSet() {
```
Comment (WibbenZ) [inline src/signalhandlingthread.cpp]: Code style

File: src/signalhandlingthread.cpp [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		//times out.
+		auto receivedSignal = sigtimedwait(&sigset, nullptr, &time);
+		if (receivedSignal != -1 && getState() != THREAD_STATE_TERMINATED) {
+			dispatchSignalHandler(receivedSignal);
+		}
+	}
+}
+
+void SignalHandlingThread::dispatchSignalHandler(int signal) {
+	switch(signal) {
```
Comment (WibbenZ) [inline src/signalhandlingthread.cpp]: Code style

File: src/signalhandlingthread.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+void SignalHandlingThread::sigcontHandler()
+{
+	//Dispatcher thread
+	std::cout << "SIGCONT received, reopening the game server..." << std::endl;
+	g_game.setGameState(GAME_STATE_NORMAL);
+}
+
+void SignalHandlingThread::sighupHandler() {
```
Comment (WibbenZ) [inline src/signalhandlingthread.cpp]: Code style

**Comment by gpedro** [comment]
:+1:

**Comment by djarek** [comment]
@marksamman 
This PR is ready. Please review it when you have the time.

**Comment by BenDol** [comment]
I'm getting `sigset_t undefined` in VS on Windows, any idea why?

**Comment by ranisalt** [comment]
> I'm getting sigset_t undefined in VS on Windows, any idea why?

Yes, probably VS is not honoring `#ifndef _WIN32` anymore. Just add `#define _WIN32` on definitions.h

**Comment by marksamman** [comment]
The new cpp file should also be added to vc14/theforgottenserver.vcxproj

**Comment by kornholi** [comment]
I don't think abusing signals is the way to approach this problem

**Comment by ranisalt** [comment]
@kornholi could you elaborate?

**Comment by djarek** [comment]
@marksamman 
Ported the signal handling code to asio::signal_set. This should make this change a little cleaner and according to asio docs should also work on Windows. Currently supported signals:
- SIGINT, SIGTERM (save game state and shutdown cleanly)
- SIGHUP (reload configs)
- SIGUSR1 (save game state)

**Comment by marksamman** [comment]
Awesome! Could you also add signals.cpp to vc14/theforgottenserver.vcxproj?

**Comment by djarek** [comment]
@marksamman 
Added the new source files to the VS project and squashed the commits.

**Comment by marksamman** [comment]
Thanks!

File: vc14/theforgottenserver.vcxproj [hunk 4]
```diff
@@ -273,6 +274,7 @@
     <ClInclude Include="..\src\scheduler.h" />
     <ClInclude Include="..\src\scriptmanager.h" />
     <ClInclude Include="..\src\server.h" />
+	<ClInclude Include="..\src\signals.h" />
```
Comment (djarek) [inline vc14/theforgottenserver.vcxproj]: Fixed.

**Comment by Mkalo** [comment]
@djarek Did you compile it on windows? I'm getting this when compiling it:
sig

I did manage to compile it placing: 
`#define SIGUSR1         30
#define SIGHUP          1`
In signal.h but I don't know if those values are correct.

**Comment by kornholi** [comment]
There's no SIGUSR1 or SIGHUP on Windows.

**Comment by Mkalo** [comment]
Defining SIGUSR1 and SIGHUP made it compile but it doesn't open. It just crashes.

**Comment by marksamman** [comment]
They should either be removed or be wrapped in #ifndef WIN32, could you open a new issue about this so it can be tracked?

**Pull request merged**